src/Domain/CoordinateMaps/BulgedCube.cpp

```diff
@@ -0,0 +1,371 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/CoordinateMaps/BulgedCube.hpp"
+
+#include <pup.h>
+
+#include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/RootFinding/RootFinder.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace {
+template <typename DType>
```

Comment:
@wthrowe @kidder and I figured something like what I have below would be better. Less code. You'll have to wait for the get_element and get_size PR to be merged.

```cpp
template <typename DType>
class RootFunction {
 public:
  RootFunction(const double radius, const double sphericity,
               const DType& physical_r_squared, const DType& x_sq,
               const DType& y_sq, const DType& z_sq) noexcept
      : radius_(radius),
        sphericity_(sphericity),
        physical_r_squared_(physical_r_squared),
        x_sq_(x_sq),
        y_sq_(y_sq),
        z_sq_(z_sq) {}

  double operator()(const double rho, const size_t i = 0) const {
    if (equal_within_roundoff(get_element(physical_r_squared_, i), 0.0)) {
      const double x_sq_over_r_sq =
          get_element(x_sq_, i) / get_element(physical_r_squared_, i);
      const double y_sq_over_r_sq =
          get_element(y_sq_, i) / get_element(physical_r_squared_, i);
      const double z_sq_over_r_sq =
          get_element(z_sq_, i) / get_element(physical_r_squared_, i);
      return sqrt(get_element(physical_r_squared_, i)) -
             radius_ * rho *
                 (1.0 / sqrt(3.0) +
                  sphericity_ *
                      (1.0 / sqrt(1.0 + square(rho) *
                                            (x_sq_over_r_sq + y_sq_over_r_sq)) +
                       1.0 / sqrt(1.0 + square(rho) *
                                            (x_sq_over_r_sq + z_sq_over_r_sq)) +
                       1.0 / sqrt(1.0 + square(rho) *
                                            (y_sq_over_r_sq + z_sq_over_r_sq)) -
                       1.0 / sqrt(2.0 + square(rho) * x_sq_over_r_sq) -
                       1.0 / sqrt(2.0 + square(rho) * y_sq_over_r_sq) -
                       1.0 / sqrt(2.0 + square(rho) * z_sq_over_r_sq)));
    } else {
      return 0.0;
    }
  }


 private:
  const double radius_;
  const double sphericity_;
  const DType& physical_r_squared_;
  const DType& x_sq_;
  const DType& y_sq_;
  const DType& z_sq_;
};
```